### PR TITLE
[MPS] Make several ops unranked to avoid graph recompilation

### PR DESCRIPTION
**Pull request by razarmehr**
- Made `bmm`, `cat`, `softmax`, `copy_cast` and `index_select` unranked. This improves the performance of LLMs and other networks, because these graphs would be invariant to shape changes, and therefore, they won't recompile. 
- Binary ops: Flatten tensors bigger than 4D to 1D, and reuse existing buffer for in-place operations (for offset 0).
- Cache the `MPSGraphExecutable` in `MPSCachedGraph`. These executables have the typeInference disabled which should help with recompilation issues on MPSGraph.
- Handle transposes in second batch of matrices in `bmm`. This helps with performance in case the second batch (i.e., `other` argument) was transposed (like the case with `QxK.t()` in Transformers).

cc @kulinseth @albanD @malfet @DenisVieriu97 @abhudev

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -2990,7 +2990,9 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
                skips=(
                    DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_grad'),
                    DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_gradgrad'),
-                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),)
+                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),
+                   # Fails with channels last test on MPS backend
+                   DecorateInfo(unittest.expectedFailure, "TestModule", "test_memory_format", device_type='mps'),)
```
Comment (albanD) [inline torch/testing/_internal/common_modules.py:2995]: Why is this new skip introduced? This used to work and doesn't anymore?

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -3265,7 +3267,9 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
     ModuleInfo(torch.nn.Tanhshrink,
                module_inputs_func=module_inputs_torch_nn_Tanhshrink,
                skips=(
-                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),)
+                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),
+                   # Fails with channels last test on MPS backend
+                   DecorateInfo(unittest.expectedFailure, "TestModule", "test_memory_format", device_type='mps'),)
```
Comment (albanD) [inline torch/testing/_internal/common_modules.py:3272]: Same as above?

File: aten/src/ATen/native/mps/operations/BinaryOps.mm [hunk 3]
```diff
@@ -82,16 +87,17 @@ static void binaryOpTensor(const Tensor& self,
...
-    output = at::empty(output_.sizes(), output_.scalar_type(), c10::nullopt, kMPS, c10::nullopt, c10::nullopt);
-    needsCopyToOutput = true;
+  // determine if this is an in-place operation
+  if (self.is_alias_of(output_) || other.is_alias_of(output_)) {
+    if (output_.storage_offset() || !output_.is_contiguous()) {
+      output = at::empty(output_.sizes(), output_.scalar_type(), c10::nullopt, kMPS, c10::nullopt, c10::nullopt);
+      needsCopyToOutput = true;
+    }
+  } else if (!output_.is_contiguous()) {
+    output_.unsafeGetTensorImpl()->empty_tensor_restride(MemoryFormat::Contiguous);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/BinaryOps.mm:100]: This is a no-op, isn't it?

File: test/test_mps.py [hunk 4]
```diff
@@ -139,8 +139,6 @@ def mps_ops_grad_modifier(ops):
 
     MACOS_BEFORE_13_3_XFAILLIST_GRAD = {
         # Failures due to precision issues (due to fast-math). These has been fixed in MacOS 13.3+
-        'masked.softmin': [torch.float32],
-        'masked.softmax': [torch.float32],
```
Comment (malfet) [inline test/test_mps.py:143]: Why are you removing those two? They still fail on older MacOS, wouldn't they?

**Review by malfet — CHANGES_REQUESTED** [review]
Looks like this changes regresses behavior with channel last (likely due to the typo in BinaryOps.mm)

Also, please do not introduce new skips, as this sounds like a regression, nor remove XFAILS unless you expect them to be fixed somehow.

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (razarmehr) [reply torch/testing/_internal/common_modules.py:2995]: No, it's not a regression. This test always failed. This skip is already in place from another PR:
https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L3069 

So I'll remove this change from my PR.

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (razarmehr) [reply torch/testing/_internal/common_modules.py:3272]: Same as above. It's not a regression and pre-existed, The skip is introduced in another PR, so I'll remove this change from my PR:
https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L3409

File: test/test_mps.py (hunk 4 above)
Comment (razarmehr) [reply test/test_mps.py:151]: As the CIs on this PR show (for all MacOS versions), these tests don't fail anymore as this PR make them unrank. They now fail as "unexpected success" so they should be removed from the list.

File: aten/src/ATen/native/mps/operations/BinaryOps.mm (hunk 3 above)
Comment (razarmehr) [reply aten/src/ATen/native/mps/operations/BinaryOps.mm:103]: No, it's an optimization. Instead of performing a gather in the Placeholder on the non-contiguous tensor, we make the empty tensor contiguous with this trick. We did the same in several ops too:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mps/operations/Pooling.mm#L126

**Comment by razarmehr** [comment]
> Looks like this changes regresses behavior with channel last (likely due to the typo in BinaryOps.mm)
> 
> Also, please do not introduce new skips, as this sounds like a regression, nor remove XFAILS unless you expect them to be fixed somehow.

It's not a regression, and pre-existed. i removed those from this PR as someone else already had landed them in another PR:
https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L3409
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mps/operations/Pooling.mm#L126

**Review by kulinseth — APPROVED** [review]

**Comment by kulinseth** [comment]
> Looks like this changes regresses behavior with channel last (likely due to the typo in BinaryOps.mm)
> 

which test was regressing? 

> Also, please do not introduce new skips, as this sounds like a regression, nor remove XFAILS unless you expect them to be fixed somehow.

The softmin started passing with the unranked fix. @malfet , Are you okay with the PR now ?

**Comment by qqaatw** [comment]
@malfet @kulinseth should we revive this one?

**Comment by malfet** [comment]
Now that we have testing we should. Though I want to land few changes that re-enables more testing, as potentially breaking something for MacOS 14, before 15 is out is not good
